pinot-core/src/main/java/org/apache/pinot/core/plan/SelectionPlanNode.java

```diff
@@ -59,12 +59,15 @@ public SelectionPlanNode(IndexSegment indexSegment, QueryContext queryContext) {
             Math.min(limit, DocIdSetPlanNode.MAX_DOC_PER_CALL)).run();
         return new SelectionOnlyOperator(_indexSegment, _queryContext, expressions, transformOperator);
       } else {
-        // Selection order-by
-        if (orderByExpressions.size() == expressions.size()) {
+        if (isAllOrderByColumnsSorted(orderByExpressions)) { // no sorting needed
+          TransformOperator transformOperator =
+                  new TransformPlanNode(_indexSegment, _queryContext, expressions, DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
+          return new SelectionOrderByOperator(_indexSegment, _queryContext, expressions, transformOperator, true);
+        } else if (orderByExpressions.size() == expressions.size()) { // Selection order-by
           // All output expressions are ordered
           TransformOperator transformOperator =
-              new TransformPlanNode(_indexSegment, _queryContext, expressions, DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
-          return new SelectionOrderByOperator(_indexSegment, _queryContext, expressions, transformOperator);
+                  new TransformPlanNode(_indexSegment, _queryContext, expressions, DocIdSetPlanNode.MAX_DOC_PER_CALL).run();
```

Comment:
Please reformat the changes with [Pinot Style](https://docs.pinot.apache.org/developers/developers-and-contributors/code-setup#intellij)